text/0001-workspace-consistent-versions.md

```diff
@@ -0,0 +1,572 @@
+
+- <details close>
+  <summary>Changes the version of a dependency line-adjacent to the upgraded dependency.</summary>
+
+  Suppose the current `HEAD` commit upgrades `jest`. On `git merge`, the following conflict manifests.
+
+  ```yaml
+  # pnpm-lock.yaml
+  packages:
+
+    packages/bar:
+      specifiers:
+  <<<<<<< HEAD
+        jest: ^28.0.0-alpha.7
+        react: ^17.0.2
+        react-dom: ^17.0.2
+  =======
+        jest: ^27.5.1
+        react: ^18.0.0-rc.3
+        react-dom: ^18.0.0-rc.3
+  >>>>>>> upgrade-react
+  ```
+  </details>
+
+- <details>
+  <summary>Add or removes a dependency line-adjacent to the upgraded dependency.</summary>
+
+  Suppose the current `HEAD` commit deletes `jest` from `bar`. On `git merge`, the following conflict manifests.
+
+  ```yaml
+  # pnpm-lock.yaml
+  packages:
+
+    packages/bar:
+      specifiers:
+  <<<<<<< HEAD
+        react: ^17.0.2
+        react-dom: ^17.0.2
+      dependencies:
+        react: 17.0.2
+        react-dom: 17.0.2_react@17.0.2
+  =======
+        jest: ^27.5.1
+        react: ^18.0.0-rc.3
+        react-dom: ^18.0.0-rc.3
+      dependencies:
+        react: 18.0.0-rc.3-next-de516ca5a-20220321
+        react-dom: 18.0.0-rc.3-next-de516ca5a-20220321_1d81fe1bff57e86f30d2d405110ea7f4
+      devDependencies:
+        jest: 27.5.1
+  >>>>>>> upgrade-react
+  ```
+
+  A similar merge conflict arises if `redux` was added since it would be declared below `react`.
+
+  ```yaml
+  # pnpm-lock.yaml
+  packages:
+
+    packages/foo:
+      specifiers:
+  <<<<<<< HEAD
+        react: ^17.0.2
+        redux: ^4.1.2
+      dependencies:
+        react: 17.0.2
+        redux: 4.1.2
+  =======
+        react: ^18.0.0-rc.3
+      dependencies:
+        react: 18.0.0-rc.3-next-de516ca5a-20220321
+  >>>>>>> upgrade-react
+  ```
+  </details>
+
+- <details>
+  <summary>Adds a new declaration of the upgraded dependency.</summary>
+
+  Suppose the current `HEAD` commit adds a `react-dom` declaration to `foo`:
+
+  ```yaml
+  # pnpm-lock.yaml
+  packages:
+
+    packages/foo:
+      specifiers:
+  <<<<<<< HEAD
+        react: ^17.0.2
+        react-dom: ^17.0.2
+      dependencies:
+        react: 17.0.2
+        react-dom: 17.0.2_react@17.0.2
+  =======
+        react: ^18.0.0-rc.3
+      dependencies:
+        react: 18.0.0-rc.3-next-de516ca5a-20220321
+  >>>>>>> upgrade-react
+  ```
+  </details>
+
+As monorepos grow in workspace package count, merge conflicts become increasingly more probable. On GitHub, any merge conflict in `pnpm-lock.yaml` blocks pull requests due to lack of custom merge driver support.
+
+With first-class support for workspace consistent versions, edits on upgrades are limited to the root `package.json` and the `workspaceConsistent` portion of `pnpm-lock.yaml`. This reduces churn in `pnpm-lock.yaml`, and **prevents merge conflicts in all the cases listed above**.
```

Comment:
I don't get it, how does it reduce the nubmer of conflicts? The same conflicts will happen just in another place, in the `workspaceConsistent` section.

Reply:
The spacing in the `workspaceConsistent` section helps a lot. I took some time tonight to commit the example merges in this RFC to a GitHub repo. I added a theoretical pull request for each scenario in an imaginary world using the `workspace-consistent:` protocol.

For the PRs in this repo:

- The `main` branch has just upgraded React.
- The `main-workspace-consistent` branch has also just upgraded React, but is using the `workspace-consistent:` protocol.

Note that all PRs using `workspace-consistent:` show as ready to merge in GitHub, while the commits not have merge conflicts in `pnpm-lock.yaml`.

|Example|Not Using the New Protocol|Using the New Protocol|
|-|-|-|
|Changes the version of a dependency line-adjacent to the upgraded dependency.|https://github.com/gluxon/pnpm-workspace-consistent-examples/pull/4|https://github.com/gluxon/pnpm-workspace-consistent-examples/pull/6|
|Add or removes a dependency line-adjacent to the upgraded dependency.|https://github.com/gluxon/pnpm-workspace-consistent-examples/pull/3|https://github.com/gluxon/pnpm-workspace-consistent-examples/pull/2|
|Adds a new declaration of the upgraded dependency.|https://github.com/gluxon/pnpm-workspace-consistent-examples/pull/5|https://github.com/gluxon/pnpm-workspace-consistent-examples/pull/1

Reply:
Looks like simply adding new lines also solves the issue: https://github.com/gluxon/pnpm-workspace-consistent-examples/pull/7

Reply:
Thank you for testing that alternative and adding an example. Mentioned in a related thread https://github.com/pnpm/rfcs/pull/1#discussion_r838156141, but that does solve 1 and 2.

I am still interested in this RFC since:

- It addresses the 3rd "_Adds a new declaration of the upgraded dependency._" case that the spacing doesn't.
- Allows `package.json` files to be more merge conflict resistant too without adding newlines to the `dependencies` section. I could see slow adoption for adding newlines to `dependencies` as a best practice in repos with consistent versions. Whereas if repos use this feature, they get the merge conflict resistance for free.
- Having `workspace-consistent:` in `package.json` still better reflects intent.
  - Namely, the "How do I add a new dependency?" is common question amongst new developers on my team. The current answer is you go check if another `package.json` has it listed and copy the version range if it does exists. That feels a bit strange. Having `pnpm add <pkg>` simply added `"<pkg>": "workspace-consistent:"` if `<pkg>` is configured feels much better.